Change heuristic-based scheduling to a 2-phase one for exchange client

highker opened the pull request:
Exchange client schedules HTTP buffer clients to fetch pages based on
heuristics. This works under the assumption that the data transfering is
under a rather stable and evenly distributed way for all the sinks.
There could be cases where the coming pages are way larger than the
previous ones. This can cause a sudden increase in the buffer size. In
the worst case, if the cluster is large enough (e.g. > 300) and has
large number of concurrent running queries (e.g. > 40), this can lead
to a worst case of 300 X 40 X 16MB = 192GB memory usage. This patch
splits the fetching protocol into a two-phase manner by first retrieving
the page sizes, reversing memory on the source, then fetching the pages
within the memory limit.

highker commented:
Here is a simple benchmark to show:
(1) the patch is not going to make a regression, and
(2) can limit the buffer size to the buffer capacity


The query is shown as follows. `checksum` is to verify the correctness of the patch and `sequence(1, 2000)` is to create large pages. We run the same query on a cluster with 10 workers with and without the patch.
```
select
checksum(w.query_id) as id,
checksum(w.bytes) as bytes,
checksum(w.rows) as rows,
checksum(n.peak_memory_bytes) as bytes,
checksum(n.total_split_cpu_time_ms) as cpu_ms,
checksum(n.seq) as seq
from
(
  select sequence(1, 2000) as seq, query_id, peak_memory_bytes, total_split_cpu_time_ms
  from nectar_presto_query_stats_r where ds='2017-05-07'
) n
JOIN
(
  select query_id, sum(bytes) as bytes, sum(rows) as rows
  from presto_hive_write where ds='2017-05-07'
  group by query_id
) w
ON w.query_id = n.query_id;
```


checksums for both runs are the same:
```
           id            |          bytes          |          rows           |          bytes          |         cpu_ms          |           seq           
-------------------------+-------------------------+-------------------------+-------------------------+-------------------------+-------------------------
 af 06 98 1d 73 15 43 63 | a4 c0 8f 46 5e e0 97 59 | aa 8c 6a c9 76 15 b2 e0 | 60 c5 5e ab 51 0e 2d 96 | 20 18 69 95 bd 5d 6f ac | 25 07 e7 5c 89 8f 7f dc
```


Each machine of the cluster has a default buffer limit of 32MB. After the change the buffer can be limited close enough to 32 MB. Given this is a cluster with only 10 workers. A larger fluctuation (> 108.5 MB buffered pages) can be observed on larger clusters without the change.
```
           query id                      max buffered   cpu time    wall time   peak memory
before     20170509_200444_00002_v4qmp   108.5 MB       21.56 min   1.20 min    53.90 MB
after      20170509_194801_00006_h3xnh   33.4 MB        18.00 min   1.19 min    53.90 MB
```

dain commented on presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java:
```diff
@@ -290,24 +294,27 @@ public synchronized void scheduleRequestIfNecessary()
+            client.scheduleRequest(new DataSize(maxBytes, BYTE));
+        }
+
+        // drain all the clients that do not have a pending page
```
I don't understand why we are doing this.  Can you comment on the purpose of this section.

highker replied:
Consider the following queue with each number denoting the pending pages (in bytes) for each client.
Suppose `availableBytes` is 8.

```
   -------------------
<- 0 0 1 2 5 6 0 0 5 0  <- enqueue
   -------------------
```

The previous loop will allow the first 5 clients to fetch pages till the quota reaches 8:


```
   -------------------
<- 6 0 0 5 0  <- enqueue
   -------------------
```

But indeed those clients without pending pages should also send requests to the server to ask if there is any page available.
So, what this loop does is to drain the queue to be:

```
   -------------------
<- 6 5  <- enqueue
   -------------------
```

Those clients are not going to fetch real data; so it is safe doing so.
Also, those requests are long poll to prevent adding them back immediately.


Will add a comment for this.

dain commented on presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java:
```diff
@@ -89,6 +92,7 @@
     private long successfulRequests;
     @GuardedBy("this")
     private long averageBytesPerRequest;
+    private final AtomicLong reservedBytes = new AtomicLong();
```
Maybe add stats for this.

dain commented on presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java:
```diff
@@ -89,6 +92,7 @@
     private long successfulRequests;
     @GuardedBy("this")
     private long averageBytesPerRequest;
```
Is this still needed?

highker commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/BufferSize.java:
```diff
@@ -0,0 +1,107 @@
+    private final String taskInstanceId;
+    private final long token;
+    private final boolean bufferComplete;
+    private final List<DataSize> pageSizes;
```
Will change some other places (e.g., pendingRead, processRead, get() etc) from DataSize into long to avoid coverting long and DataSize back and forth

highker commented on presto-main/src/main/java/com/facebook/presto/operator/HttpPageBufferClient.java:
```diff
@@ -343,6 +366,16 @@ public void onSuccess(PagesResponse result)
 
                         if (token == HttpPageBufferClient.this.token) {
                             pages = result.getPages();
+                            if (pages.size() > pendingPages.size()) {
```
This means we have fetched more pages than what we know. For example, suppose a client has 3 pending pages, then the exchange client should only give the client 3 pending pages worth of quota even if the server may have 5 pages. If the client comes back with 5 pages, there is definitely something wrong with the scheduler.

highker commented on presto-main/src/main/java/com/facebook/presto/operator/HttpPageBufferClient.java:
```diff
@@ -287,36 +300,46 @@ public synchronized void scheduleRequest()
     }
 
-    private synchronized void initiateRequest()
+    private synchronized void initiateRequest(DataSize maxSize)
```
This should be `synchronized` right? The function is invoked by an executor with a delay.

dain replied:
Yes... missed that

highker commented on presto-main/src/main/java/com/facebook/presto/operator/HttpPageBufferClient.java:
```diff
@@ -343,6 +366,16 @@ public void onSuccess(PagesResponse result)
+                            }
+                            for (int i = 0; i < pages.size(); i++) {
+                                if (pages.get(i).getRetainedSizeInBytes() > pendingPages.get(i).toBytes()) {
+                                    // TODO: why after SerDe the size sometimes becomes smaller?
```
Checked `PagesSerde`. When serializing and compressing a page on the server side, we construct the page with a `DynamicSliceOutput`. The retained size could be larger than the actual size unless we do a compression. A compression will make a copy of the dynamic slice with the actual length. After we transfer the page to the client side, the retained size is the actual size. Given this property, the client may reserve more memory than needed. This possibly won't be a big issue; the memory is still bounded by the buffer size. It is also impossible that a client reserves huge amount of memory and only fetches a page in a small size. Because if that is the case, the serve should compress the page. A compressed page should have the same retained size on both server and client.

Will add a comment.

highker commented:
Thanks @dain for the review; major revisions:
- split `getXXX()` into `getXXXSummary()` and `getXXXData()`
- split rest api `v1/{task}/results/{buffer}/{token}` into `v1/{task}/buffer/{buffer}/[summary|data]/{token}`
- add more stats to `httpBufferClient` and `exchangeClient` for debugging
- all comments are addressed

dain commented on presto-main/src/main/java/com/facebook/presto/execution/SqlTaskManager.java:
```diff
@@ -338,24 +337,25 @@ public TaskInfo updateTask(Session session, TaskId taskId, Optional<PlanFragment
-                    return tasks.getUnchecked(taskId).getTaskData(bufferId, startingSequenceId, result.getPageSizesInBytes().stream().mapToLong(Long::longValue).sum());
-                });
+        checkArgument(startingSequenceId >= 0, "startingSequenceId is negative");
+        requireNonNull(maxBytes, "maxSize is null");
```
typo `maxSize`

highker commented:
verifier result looks good
```
com.facebook.presto.verifier.Verifier	Progress: 3227 valid, 0 failed, 432 skipped, 100.00% done
com.facebook.presto.verifier.Verifier	Results: 3227 / 0 (432 skipped)
```

highker commented:
Blocked by https://github.com/prestodb/presto/pull/8594; this patch requires http2 to be turned on

highker commented:
Some follow up actions from #10030:
- Do we need to revert #10030 once this PR is landed? Original I was thinking yes; but seems not. Because the new protocol is not going to send `getSummary` to acknowledge the received pages if the client's processing pages is slower than the servers' creating pages.
- For arbitrary buffer, the reservation/assignment mechanism is still there. It is possible that pages are reserved by one client but it takes a long time for that client to come back and fetch those pages. At the mean time, other clients may have nothing to reserve.

highker commented:
hand over to @wenleix to finish up
